File: 2023/day01/go/src/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"unicode"
)

func convert(input string) int {

	if len(input) == 0 {
		return -1
	}

	// fmt.Print("input: ", input, " ")
	if unicode.IsDigit(rune(input[0])) {
		x, _ := strconv.Atoi(string(input[0]))
		// fmt.Println("digit", x)
		return x
	}

	if strings.HasPrefix(input, "zero") {
		return 0
	} else if strings.HasPrefix(input, "one") {
		return 1
	} else if strings.HasPrefix(input, "two") {
		return 2
	} else if strings.HasPrefix(input, "three") {
		return 3
	} else if strings.HasPrefix(input, "four") {
		return 4
	} else if strings.HasPrefix(input, "five") {
		return 5
	} else if strings.HasPrefix(input, "six") {
		return 6
	} else if strings.HasPrefix(input, "seven") {
		return 7
	} else if strings.HasPrefix(input, "eight") {
		return 8
	} else if strings.HasPrefix(input, "nine") {
		return 9
	} else {
		return -1
	}
}

func part1(input string) int {
	sum := 0
	for _, line := range strings.Split(input, "\n") {
		digits := ""
		for _, c := range line {
			if unicode.IsDigit(c) {
				digits += string(c)
			}
		}
		if len(digits) == 0 {
			continue
		}

		first, err := strconv.Atoi(string(digits[0]))
		if err != nil {
			panic(err)
		}
		last, err := strconv.Atoi(string(digits[len(digits)-1]))
		if err != nil {
			panic(err)
		}
		sum += 10*first + last
	}
	return sum
}

func part2(input string) int {
	sum := 0
	for _, line := range strings.Split(input, "\n") {
		digits := []int{}

		for i := range line {
			value := convert(line[i:])
			if value != -1 {
				digits = append(digits, value)
			}
		}
		sum += 10*digits[0] + digits[len(digits)-1]
	}
	return sum

}

func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}

	input := string(data)

	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}

File: 2023/day01/cpp/main.cpp
#include <fstream>
#include <iostream>
#include <iterator>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

int part_1(string input) {
  string line;
  auto sum = 0;
  stringstream ss(input);

  while (getline(ss, line, '\n')) {
    string digits;
    copy_if(line.begin(), line.end(), back_inserter(digits),
            [](char c) { return std::isdigit(c); });

    string number;

    if (digits.empty()) continue;

    number.push_back(digits.front());
    number.push_back(digits.back());
    sum += stoi(number);
  }
  return sum;
}

int convert(string input) {
  static const std::unordered_map<std::string, int> digitMap = {
      {"zero", 0}, {"one", 1}, {"two", 2},   {"three", 3}, {"four", 4},
      {"five", 5}, {"six", 6}, {"seven", 7}, {"eight", 8}, {"nine", 9}};

  if (input.empty()) return -1;
  if (isdigit(input[0])) return input[0] - '0';

  for (auto const& [key, val] : digitMap)
    if (input.substr(0, key.size()) == key) return val;

  return -1;
}

int part_2(string input) {
  string line;
  auto sum = 0;
  stringstream ss(input);
  int count = 0;

  while (getline(ss, line, '\n')) {
    count++;
    vector<int> digits;

    for (int i = 0; i < line.size(); ++i) {
      int con = convert(line.substr(i, line.size()));
      if (con != -1) digits.push_back(con);
    }

    if (digits.size() == 0) continue;
    sum += 10 * digits.front() + digits.back();
  }

  return sum;
}

int main() {
  ifstream t("../input.txt");
  stringstream input;
  input << t.rdbuf();

  cout << "Part 1: " << part_1(input.str()) << endl;
  cout << "Part 2: " << part_2(input.str()) << endl;

  return 0;
}
